web-modules/javax-servlets-2/src/main/java/com/baeldung/servlets/FileUploadServlet.java

```diff
@@ -0,0 +1,64 @@
+package com.baeldung.servlets;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.MultipartConfig;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.Part;
+
+@WebServlet(name = "FileUploadServlet", urlPatterns = "/upload")
+@MultipartConfig(fileSizeThreshold = 1024 * 1024, maxFileSize = 1024 * 1024 * 5, maxRequestSize = 1024 * 1024 * 5 * 5)
+public class FileUploadServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+    private static final String UPLOAD_DIRECTORY = "uploads";
+
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+
+        String uploadPath = getServletContext().getRealPath("") + File.separator + UPLOAD_DIRECTORY;
+        File uploadDir = new File(uploadPath);
+        if (!uploadDir.exists()) {
+            uploadDir.mkdir();
+        }
+
+        Part filePart = request.getPart("file");
+        if (filePart != null) {
+            String fileName = Paths.get(filePart.getSubmittedFileName()).getFileName().toString();
+            if(fileName.isEmpty()){
+                response.getWriter().println("Invalid File Name!");
+                return;
+            }
+            if(!fileName.endsWith(".txt")){
+                response.getWriter().println("Only .txt files are allowed!");
+                return;
+            }
+
+            File file = new File(uploadPath, fileName);
+
+            try (InputStream fileContent = filePart.getInputStream()) {
+                Files.copy(fileContent, file.toPath(), StandardCopyOption.REPLACE_EXISTING);
```

Comment:
This is vulnerable to directory traversal. I recommend linking to https://cheatsheetseries.owasp.org/cheatsheets/File_Upload_Cheat_Sheet.html